### PR TITLE
GH3226: Resolve neutral culture assembly if .resources file is not found.

### LINKED ISSUE #3226: Still can't resolve resource assemblies after GH2734
This issue is continuation of [GH2734](https://github.com/cake-build/cake/issues/2734).

It's like making 5 mistakes in a 3-letter catch... Seems no one actually tested or reviewed this properly.

1. There is some mess between using full and short assembly name in `ScriptAssemblyResolver`. For example `Assembly.Load()` method expects full name not a short one. Also see point 4 below.
2. Assembly version is ignored. This leads to possible compatibility issues when different versions of resolved assembly have breaking changes.
3. `try-catch` scope is too wide. If `.resources` assembly can't be loaded the `Assembly.Load()` throws a `FileNotFoundException`. Due to `.resources` assembly name check is inside `try` block after `Assembly.Load()` call it is never executed for `.resources` assembly. Additionally recursive call to `AssemblyResolve()` must be outside of `try` block since it handles exceptions on its own.
4. Finally the actual `.resources` assembly name check is always `false` because it is performed against short assembly name which never contains comma separators or spaces :)

### PR DESCRIPTION
**Pull request by mmarinchenko**
Fixes #3226.

### CODE REVIEW

**Comment by dnfadmin** [comment]
[CLA assistant check](https://cla.dotnetfoundation.org/cake-build/cake?pullRequest=3227) <br/>Thank you for your submission, we really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla.dotnetfoundation.org/cake-build/cake?pullRequest=3227) before we can accept your contribution.<br/><br/>:x: mmarinchenko [sign now](https://cla.dotnetfoundation.org/cake-build/cake?pullRequest=3227)<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla.dotnetfoundation.org/check/cake-build/cake?pullRequest=3227) it.</sub>

**Comment by mmarinchenko** [comment]
Sample diagnostic output after fix:
```
Resolving assembly 'NuGet.Protocol.resources, Version=5.8.1.7021, Culture=ru-RU, PublicKeyToken=31bf3856ad364e35'...
Exception occurred while resolving assembly NuGet.Protocol.resources: Could not load file or assembly 'NuGet.Protocol.resources, Version=5.8.1.7021, Culture=ru-RU, PublicKeyToken=31bf3856ad364e35'.
Trying to resolve assembly NuGet.Protocol.resources as 'NuGet.Protocol, Version=5.8.1.7021, Culture=ru-RU, PublicKeyToken=31bf3856ad364e35'...
Assembly NuGet.Protocol resolved as 'NuGet.Protocol, Version=5.8.1.7021, Culture=neutral, PublicKeyToken=31bf3856ad364e35'
```

**Comment by augustoproiete** [comment]
Thanks for the PR @mmarinchenko.

Could you take a look at the integration tests failing on Ubuntu?

https://github.com/cake-build/cake/runs/2023599068?check_suite_focus=true

```
Error: Could not find file 'System.Private.Uri.resources'.
An error occurred when executing task 'Cake.Core.Scripting.AddinDirective.LoadNetStandardAddin'.
Error: One or more errors occurred. (.NET Core CLI: Process returned an error (exit code 1).)
.NET Core CLI: Process returned an error (exit code 1).
While testing: /home/runner/work/cake/cake/tests/integration/tools/Cake/Cake.exe
Cake.Core.CakeException: Cake: Process returned an error (exit code 1).
```

**Comment by mmarinchenko** [comment]
@augustoproiete Hi!

> Could you take a look at the integration tests failing on Ubuntu?

Unfortunately it's not clear for me why this happens on ubuntu. I need a time to investigate this.

**Comment by augustoproiete** [comment]
@mmarinchenko No worries! I'll mark the PR as draft for now and just ping us when you think it's ready for review.

**Comment by mmarinchenko** [comment]
@augustoproiete Ok, thanks!

**Comment by mmarinchenko** [comment]
@augustoproiete 

I guess I found the reason why integration tests failed on Ubuntu.

Loading of satellite assemblies (i.e. `*.resources`) on behalf of main one is broken on versions before **.NET Core 3.0** (see https://github.com/dotnet/coreclr/pull/24191 for details). So I added additional check for this.

Now main assembly is not resolved instead of satellites if Cake is running on .NET Core runtime older than 3.0. In your test environment .NET Core runtime **v2.1.25** is used as a "system" one for tooling on Ubuntu - that's why it all happened)

**Review by devlead — CHANGES_REQUESTED** [review]
On MacOS these tests fail, been able to repro same failure on my Mac, probably something unrelated on my mac.

File: src/Cake/Infrastructure/Scripting/ScriptAssemblyResolver.cs [hunk 1]
```diff
@@ -28,46 +29,74 @@ public void Dispose()
...
+            _log.Verbose($"Resolving assembly '{fullName}' using runtime installed at '{RuntimeEnvironment.GetRuntimeDirectory()}'...");
+            return AssemblyResolve(assemblyName);
         }
 
-        private Assembly AssemblyResolve(AssemblyName name)
+        private Assembly AssemblyResolve(AssemblyName assemblyName)
         {
+            var fullName = assemblyName.FullName;
+            var shortName = assemblyName.Name;
+            var version = assemblyName.Version.ToString();
```
Comment (devlead) [inline src/Cake/Infrastructure/Scripting/ScriptAssemblyResolver.cs:56]: Why the ToString() here? Version has a comparer 

So `x.GetName().Version == version` should work?

File: src/Cake/Infrastructure/Scripting/ScriptAssemblyResolver.cs [hunk 2]
```diff
@@ -28,46 +29,74 @@ public void Dispose()
...
+#if NETCORE
+            // Since .NET Core 3.0
+            var runtimeMainVer = Environment.Version.Major;
+            if (runtimeMainVer >= 5
+                || (runtimeMainVer == 3
+                    && RuntimeInformation.FrameworkDescription.StartsWith(".NET Core")))
+#endif
             {
-                _log.Verbose($"Exception while resolving assembly {name.Name}: {ex.Message}");
+                var resources = ".resources";
```
Comment (devlead) [inline src/Cake/Infrastructure/Scripting/ScriptAssemblyResolver.cs:86]: This can be a constant

**Comment by mmarinchenko** [comment]
@devlead 

I don't have access to logs on bitrise. And I don't have a Mac, sorry)

I can see all tests are passed on [GitHub Build / Build (macos-latest)](https://github.com/cake-build/cake/pull/3227/checks?check_run_id=2668451659). What's the difference between these configurations?

**Review by devlead — CHANGES_REQUESTED** [review]

**Comment by devlead** [comment]
> @devlead
> 
> I don't have access to logs on bitrise. And I don't have a Mac, sorry)
> 
> I can see all tests are passed on [GitHub Build / Build (macos-latest)](https://github.com/cake-build/cake/pull/3227/checks?check_run_id=2668451659). What's the difference between these configurations?

Log from Bitrise
[4c49a72ec940c49e.log](https://github.com/cake-build/cake/files/6542053/4c49a72ec940c49e.log)
```
[xUnit.net 00:00:04.22]     [FATAL ERROR] System.AppDomainUnloadedException
```
Could be differences in Mono version or MacOs version installed.
But also maybe something temporary I'll rerun CI.

**Comment by devlead** [comment]
> > @devlead
> > 
> > I don't have access to logs on bitrise. And I don't have a Mac, sorry)
> > 
> > I can see all tests are passed on [GitHub Build / Build (macos-latest)](https://github.com/cake-build/cake/pull/3227/checks?check_run_id=2668451659). What's the difference between these configurations?
> 
> Log from Bitrise
> [4c49a72ec940c49e.log](https://github.com/cake-build/cake/files/6542053/4c49a72ec940c49e.log)
> ```
> [xUnit.net 00:00:04.22]     [FATAL ERROR] System.AppDomainUnloadedException
> ```
> Could be differences in Mono version or MacOs version installed.
> But also maybe something temporary I'll rerun CI.
> 

Failed test seems to been a temporary thing.

File: src/Cake/Infrastructure/Scripting/ScriptAssemblyResolver.cs (hunk 1 above)
Comment (mmarinchenko) [reply src/Cake/Infrastructure/Scripting/ScriptAssemblyResolver.cs]: Yes, `.ToString()` is unnecessary here.

File: src/Cake/Infrastructure/Scripting/ScriptAssemblyResolver.cs (hunk 2 above)
Comment (mmarinchenko) [reply src/Cake/Infrastructure/Scripting/ScriptAssemblyResolver.cs]: Done.

File: src/Cake/Infrastructure/Scripting/ScriptAssemblyResolver.cs [hunk 3]
```diff
@@ -28,46 +29,74 @@ public void Dispose()
...
-                    }
-                }
+            if (assembly != null)
+            {
+                _log.Verbose($"Assembly {shortName} resolved as '{assembly.FullName}' (file location: '{assembly.Location}')");
                 return assembly;
             }
-            catch (Exception ex)
+
+#if NETCORE
```
[the remark this answers is not in the document]
Comment (mmarinchenko) [reply src/Cake/Infrastructure/Scripting/ScriptAssemblyResolver.cs]: Oh, good point, thanks! In this case conditional compilation is also not necessary here. Fixed it.

**Review by devlead — APPROVED** [review]
LGTM 👍

**Comment by devlead** [comment]
@mmarinchenko your changes have been merged, thanks for your contribution 👍

**Pull request merged**